apps-rendering/src/newsletter.ts

```diff
@@ -0,0 +1,254 @@
+	const middle = Math.floor(numberOfElements / 2);
+	return [Math.max(0, middle - 3), Math.min(numberOfElements, middle + 3)];
+}
+
+/**
+ *  Find the place to insert a insert a sign-up block in to an article body,
+ *  using these rules:
+ *	 - Don't put straight after bold text
+ *	 - Don't put under headings
+ *	 - Must have plain body text above and below
+ *   - Must be within 3 elements from the middle of the article
+ *   - Should not be next to an adSlot (currented addressed by the
+ *     existing adslot placement module - the sidgn-up block is enclosed
+ *     in an <aside> and the adslots are placed after <p> elements  )
+ *   - The best place is the last place meeting the criteria above.
+ *
+ * @param body an Item.Body
+ * @returns A Result with the insertIndex or an error
+ */
+function findInsertIndex(body: Body): Result<string, number> {
+	const positionsWithContent: ElementCategoryAndIndex[] = body
+		.map((element, index) => ({
+			index,
+			category: categoriseElement(element),
+		}))
+		.filter(
+			(item) =>
+				![ElementCategory.Error, ElementCategory.WhiteSpace].includes(
+					item.category,
+				),
+		);
+
+	// get the slice from positionsWithContent that has the elements
+	// with the target range.
+	// The slice starts are minIndex-1 as the isSuitable function
+	// needs to check the category of the element before the element
+	// it is testing.
+	const [minIndex, maxIndex] = getRange(positionsWithContent.length);
+	const possibleElementsToPlaceBefore = positionsWithContent.slice(
+		minIndex - 1,
+		maxIndex,
+	);
+
+	function isSuitable(index: number): boolean {
+		const elementBeforeInsertPoint =
+			possibleElementsToPlaceBefore[index - 1];
+		const elementAfterInsertPoint = possibleElementsToPlaceBefore[index];
+
+		// NOTE - this is a type safety check. TypeScript will cast a value
+		// accessed from an array index as being of type of the Array member, even
+		// if the index is out of bounds and the value is actually undefined.
+		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- because 👆
+		if (!elementBeforeInsertPoint || !elementAfterInsertPoint) {
+			return false;
+		}
+
+		return (
+			// the element before must be regular text (not bold paragraph)
+			[ElementCategory.ParagraphText].includes(
+				elementBeforeInsertPoint.category,
+			) &&
+			// the element after must be regular text, or a bold paragraph.
+			[
+				ElementCategory.ParagraphText,
+				ElementCategory.BoldParagraphText,
+			].includes(elementAfterInsertPoint.category)
+		);
+	}
+
+	// Define a recursive function go backwards through the possible
+	// positions until it finds one that is suitable or reaches
+	// possibleElementsToPlaceBefore[0] - which is actually out of range, but
+	// has to be include to test if possiblePoistions[1] is suitable.
+	const findLastSuitablePosition = (
+		idx: number,
+	): Result<string, ElementCategoryAndIndex> => {
+		if (idx <= 0) {
```

Comment:
I think this could end up searching all the way back to the beginning of the body, which might be a bug?

Reply:
That makes sense! I hadn't realised the array was sliced beforehand

Reply:
`idx` is the index within `possibleElementsToPlaceBefore`, the array of `ElementCategoryAndIndex` , which is already bound by `minIndex` and `maxIndex` - (IE it's `possibleElementsToPlaceBefore` that is being iterated over backward, not the whole body).

I've rename the variable to be clearer which index we are using.

Also added a new test to make sure the signup block can only be placed in the middle.